.github/workflows/ci.yml

```diff
@@ -17,42 +17,66 @@ on:
       - 'examples/**'
       - '.gitignore'
 
+env:
+  GREPTIMEAI_HOST: ${{ secrets.GREPTIMEAI_HOST }}
+  GREPTIMEAI_USERNAME: ${{ secrets.GREPTIMEAI_USERNAME }}
+  GREPTIMEAI_PASSWORD: ${{ secrets.GREPTIMEAI_PASSWORD }}
+  GREPTIMEAI_DATABASE: ${{ secrets.GREPTIMEAI_DATABASE }}
+  GREPTIMEAI_TOKEN: ${{ secrets.GREPTIMEAI_TOKEN }}
+  OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+
 jobs:
-  ci:
+  ci-3-8:
     runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
 
     steps:
       - uses: actions/checkout@v4
-      - name: Set up Python ${{ matrix.python-version }}
+      - name: Set up Python 3.8
         uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: "3.8"
+      - name: Run shared steps
+        uses: ./.github/actions/
 
-      - name: Display Python version
-        run: python -c "import sys; print(sys.version)"
+  ci-3-9:
+    runs-on: ubuntu-latest
+    needs: ci-3-8
 
-      - name: Install dependencies
-        run: |
-          python -m pip install --upgrade pip
-          pip install .
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python 3.9
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.9"
+      - name: Run shared steps
+        uses: ./.github/actions/
 
-      - name: Install Dev dependencies
-        run: |
-          pip install langchain openai pytest ruff mypy
+  ci-3-10:
+    runs-on: ubuntu-latest
+    needs: ci-3-9
 
-      # TODO(yuanbohan): code coverage with pytest-cov
-      - name: Test with pytest
-        run: |
-          pytest tests/ -v
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - name: Run shared steps
+        uses: ./.github/actions/
+
+  ci-3-11:
+    runs-on: ubuntu-latest
+    needs: ci-3-10
 
-      - name: Lint & Format with Ruff
-        run: |
-          ruff check --output-format=github .
-          ruff format .
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+      - name: Run shared steps
+        uses: ./.github/actions/
 
-      - name: Static Type Checking with mypy
-        run: |
-          mypy $(git ls-files '*.py') --check-untyped-defs
+concurrency:
+  group: ${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
```

Comment:
1. refer this https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymax-parallel。 To check if this can restrict only one job is running in the matrix context
2. if `github.repository` or `github.repository_id` is the expected group name? for each PR will truncate tables, and they can not affect each other. Refer this for more context variables https://docs.github.com/en/actions/learn-github-actions/contexts#github-context